### PR TITLE
Breaking: Switch to RuleTester (fixes #3186)

### LINKED ISSUE #3186: Create RuleTester to replace ESLintTester
By moving ESLintTester into this package, we're now also bound to include chai and mocha, which sucks. It would be better to have a tester that doesn't have bindings to any particular testing frameworks. So, I'm going to make a new RuleTester object, which would be used like this:

```js
var RuleTester = require("../../../lib/testers/rule-tester");
var theRule = require("../../../lib/rules/whatever");

var tester = new RuleTester(theRule);
tester.add({
    valid: {}
    invalid: {}
});
```

Then, we'd just use `describe()` directly in our tests and I can remove ESLintTester from the package. Thus, 1.0.0 would deprecate ESLintTester in favor of RuleTester.

Thoughts?

### PR DESCRIPTION
**Pull request by nzakas**
In this PR:

1. Removed `eslint-tester`
1. Added `RuleTester`
1. Updated all rule tests.
1. Stopped exporting `validator` (no longer necessary)
1. Updated documentation

### CODE REVIEW

File: docs/developer-guide/working-with-rules.md [hunk 1]
```diff
@@ -306,13 +290,13 @@ invalid: [
 
 In this case, the message is specific to the variable being used and the AST node type is `Identifier`.
 
-Similar to the valid cases, you can also specify `args` to be passed to the rule:
+Similar to the valid cases, you can also specify `options` to be passed to the rule:
```
Comment (btmills) [inline docs/developer-guide/working-with-rules.md]: There's uneven spacing between the brackets for `options` on line 269 above.

File: docs/user-guide/migrating-to-1.0.0.md [hunk 2]
```diff
@@ -111,8 +111,32 @@ In 0.x, the `cli` object was exported for use by external tools. It was later de
 
 ## Deprecating eslint-tester
```
Comment (btmills) [inline docs/user-guide/migrating-to-1.0.0.md]: This header can be changed to "Replacing eslint-tester"

**Comment by btmills** [comment]
Saw a couple possible improvements in the docs. Code looks good - the CI failure was a 2s timeout on io.js, but everything else passed.

File: docs/user-guide/migrating-to-1.0.0.md (hunk 2 above)
Comment (nzakas) [inline docs/user-guide/migrating-to-1.0.0.md]: Well, `eslint-tester` is its own package that will still be available, so deprecating is the right term. You can still use it, but you do so at your own risk.

Comment (btmills) [inline docs/user-guide/migrating-to-1.0.0.md]: Ah, that makes sense.

**Comment by IanVS** [comment]
Gave Travis a kick...

File: tests/lib/rules/comma-dangle.js [hunk 3]
```diff
@@ -11,22 +11,22 @@
...
+ruleTester.run("comma-dangle", rule, {
     valid: [
         "var foo = { bar: 'baz' }",
         "var foo = {\nbar: 'baz'\n}",
         "var foo = [ 'baz' ]",
         "var foo = [\n'baz'\n]",
-        "[,,]",
+        "[,]",
         "[\n,\n,\n]",
         "[,]",
```
Comment (IanVS) [inline tests/lib/rules/comma-dangle.js]: With this change, lines 29 and 31 seem to be duplicates.  

Same goes for 48/50, and 64/66

File: tests/lib/rules/comma-spacing.js [hunk 4]
```diff
@@ -99,13 +99,13 @@ eslintTester.addRuleTest("lib/rules/comma-spacing", {
...
+        {code: "var arr = [,];", args: [2, {before: false, after: false}]},
+        {code: "var arr = [1,];", args: [2, {before: false, after: false}]},
         {code: "var arr = [,2,];", args: [2, {before: false, after: false}]},
-        {code: "var arr = [,,3];", args: [2, {before: false, after: false}]},
+        {code: "var arr = [,3];", args: [2, {before: false, after: false}]},
         {code: "var arr = [1,2,];", args: [2, {before: false, after: false}]},
         {code: "var arr = [,2,3];", args: [2, {before: false, after: false}]},
-        {code: "var arr = [1,,3];", args: [2, {before: false, after: false}]},
+        {code: "var arr = [1,3];", args: [2, {before: false, after: false}]},
         {code: "var arr = [1,2,3];", args: [2, {before: false, after: false}]},
```
Comment (IanVS) [inline tests/lib/rules/comma-spacing.js]: There now seem to be duplicated tests here:

- 99/103
- 100/105
- 101/108

Was this intended?

Comment (nzakas) [inline tests/lib/rules/comma-spacing.js]: No, overly aggressive find and replace. I'll fix those. Nice catch!

File: docs/developer-guide/working-with-rules.md (hunk 1 above)
Comment (nzakas) [inline docs/developer-guide/working-with-rules.md]: If only we had a markdown plugin I could use to lint code examples. ;-)

**Pull request merged**